### ISSUE TITLE
Decoding & flattening an array

### ISSUE THREAD

**Issue by eimantas** [thoughtbot/Argo#398]
I have response that I'd like to flatten:

```
{
    "foo": "bar"
    "bazCollection": [
      { "bazId": 1 },
      { "bazId": 2 },
    ]
}
```

Is there a combination of operators that I can use to `decode` this response to following struct:

```
struct FooResponse {
  let foo: String
  let bazIds: [Int]
}
```

**Comment by FranciscoAmado** [comment]
Hi @eimantas,

I believe you can do something like:
```
extension FooResponse: Decodable {

    static func decode(j: JSON) -> Decoded<FooResponse> {

        return curry(FooResponse.init)
            <^> j <| "foo"
            <*> j <|| ["bazCollection", "bazId"]
    }
}
```

**Comment by jkennington** [comment]
There is probably a cleaner way but this will do what you want.

```
import Foundation
import Argo
import Curry

var jsonString = "{\"foo\":\"bar\",\"bazCollection\":[{\"bazId\":1},{\"bazId\":2}]}"

struct FooResponse {
    let foo: String
    let bazIds: [Int]
}

extension FooResponse: Decodable {
    static func decode(json: JSON) -> Decoded<FooResponse> {
        let decodedBazIds: Decoded<[BazId]> = curry([BazId].init) <^> json <|| "bazCollection"
        let decodedIds: Decoded<[Int]> = decodedBazIds.map { $0.map({ $0.id }) }
        return curry(FooResponse.init)
            <^> json <| "foo"
            <*> decodedIds
    }
    
    struct BazId: Decodable {
        let id: Int

        static func decode(json: JSON) -> Decoded<BazId> {
            return curry(BazId.init)
                <^> json <| "bazId"
        }
    }
}

let json = try! NSJSONSerialization.JSONObjectWithData(jsonString.dataUsingEncoding(NSUTF8StringEncoding)!, options: [])
var result: Decoded<FooResponse> = decode(json)
```

**Comment by gfontenot** [comment]
Oh, interesting. This is an unusual path, and we don't have an optimized API for it. Seems like what we _really_ need is some sort of keypath support. Unfortunately we don't have anything like that right now.

This turns into another one of these interesting "follow the types" puzzles we run into a lot. The end result that I came up with is _really_ similar to the one proposed by @jkennington, but I'm preferring to use specialized functions instead of the intermediate type.

The root of the issue is that we need to go through a couple of awkward transformations of our data and it's not immediately obvious what the path should be. We start with `JSON`, and somehow need to get to `[Int]`. We know based on our expected json object that this should be true:

```swift
let collection: Decoded<[JSON]> = json <|| "bazCollection"
```

So now we're left with needing to go from `Decoded<[JSON]> -> Decoded<[Int]>`. We also know (because of our expected json) that `JSON` is an object with a `bazId` key. So our first attempt will be to map over that array, pull out that key, and decode it into an `Int`. Unfortunately, since we're dealing with a decoded array we need to do an awkward double map. We're one step closer but not quite there:

```swift
let collection: Decoded<[JSON]> = json <|| "bazCollection"
let decodedInts: Decoded<[Decoded<Int>]> = collection.map { $0.map { $0 <| "bazId" } }
```

That gets us to that weird nested `Decoded` type. If we could flatten that, it'd get us to the type we actually want.

The first step towards that goal will be to move from `[Decoded<Int>]` to `Decoded<[Int]>`. Luckily for us, that type lines up perfectly with the free function we're exporting called `sequence` (`[Decoded<T>] -> Decoded<[T]>`). So tossing that into the mix gets us to:

```swift
let collection: Decoded<[JSON]> = json <|| "bazCollection"
let decodedInts: Decoded<Decoded<[Int]>> = collection.map { sequence($0.map { $0 <| "bazId" }) }
```

And hey, now that we have two `Decoded` types like that, we should just be able to change our initial `map` to a `flatMap` and get to our final goal:

```swift
let collection: Decoded<[JSON]> = json <|| "bazCollection"
let decodedInts: Decoded<[Int]> = collection.flatMap { sequence($0.map { $0 <| "bazId" }) }
```

A little more complex than I'd like ideally, but it compiles and therefore we know it produces the correct types. So assuming our string keys are correct, it will decode what we expect it to decode.

fwiw, I dropped this into a single line helper function when I was playing with it, which I think pulls the complexity out of the main parser nicely and makes it a bit more readable:

```swift
extension FooResponse: Decodable {
  static func decode(_ json: JSON) -> Decoded<FooResponse> {
    return curry(FooResponse.init)
      <^> json <| "foo"
      <*> decodeIDs(json)
  }
}

private func decodeIDs(_ json: JSON) -> Decoded<[Int]> {
  return (json <|| "bazCollection").flatMap { sequence($0.map { $0 <| "bazId" }) }
}
```
